src/vos/vos_io.c

```diff
@@ -78,6 +78,196 @@ struct vos_io_context {
 				 ic_size_fetch:1;
 };
 
+struct dedup_entry {
+	d_list_t	 de_link;
+	uint8_t		*de_csum_buf;
+	int		 de_csum_len;
+	struct bio_iov	 de_biov;
+	int		 de_ref;
+};
+
+static inline struct dedup_entry *
+dedup_rlink2entry(d_list_t *rlink)
+{
+	return container_of(rlink, struct dedup_entry, de_link);
+}
+
+static bool
+dedup_key_cmp(struct d_hash_table *htable, d_list_t *rlink,
+	      const void *key, unsigned int ksize)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+	struct dcs_csum_info	*csum = (struct dcs_csum_info *)key;
+	unsigned int		 len;
+
+	D_ASSERT(entry->de_csum_len != 0);
+	len = csum->cs_nr * csum->cs_len;
+	D_ASSERT(len != 0);
+
+	if (entry->de_csum_len != len)
+		return false;
+
+	D_ASSERT(csum->cs_csum != NULL);
+	D_ASSERT(entry->de_csum_buf != NULL);
+
+	return memcmp(entry->de_csum_buf, csum->cs_csum, len) == 0;
+}
+
+static uint32_t
+dedup_key_hash(struct d_hash_table *htable, const void *key,
+	       unsigned int ksize)
+{
+	struct dcs_csum_info	*csum = (struct dcs_csum_info *)key;
+	unsigned int		 key_len = csum->cs_nr * csum->cs_len;
+
+	D_ASSERT(key_len != 0);
+	D_ASSERT(csum->cs_csum != NULL);
+
+	return d_hash_string_u32((const char *)csum->cs_csum, key_len);
+}
+
+static void
+dedup_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+
+	entry->de_ref++;
+}
+
+static bool
+dedup_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+
+	D_ASSERT(entry->de_ref > 0);
+	entry->de_ref--;
+
+	return entry->de_ref == 0;
+}
+
+static void
+dedup_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
```

Comment:
(style)  'dedup' may be misspelled - perhaps 'dedupe'?
(style)  'dedup' may be misspelled - perhaps 'dedupe'?